alf/algorithms/ppo_loss.py

```diff
@@ -98,58 +99,19 @@ def __init__(self,
-            training_info.collect_action_distribution, training_info.action,
-            self._action_spec)
-        sample_action_log_probs = tf.stop_gradient(sample_action_log_probs)
-
-        action_log_prob = tfa_common.log_probability(
-            current_policy_distribution, training_info.action,
-            self._action_spec)
-        if self._log_prob_clipping > 0.0:
-            action_log_prob = tf.clip_by_value(action_log_prob,
-                                               -self._log_prob_clipping,
-                                               self._log_prob_clipping)
-        if self._check_numerics:
-            action_log_prob = tf.debugging.check_numerics(
-                action_log_prob, 'action_log_prob')
-
-        # Prepare both clipped and unclipped importance ratios.
-        importance_ratio = tf.exp(action_log_prob - sample_action_log_probs)
-        if self._check_numerics:
-            importance_ratio = tf.debugging.check_numerics(
-                importance_ratio, 'importance_ratio')
-        importance_ratio_clipped = tf.clip_by_value(
-            importance_ratio, 1 - self._importance_ratio_clipping,
-            1 + self._importance_ratio_clipping)
-
+        name_scope = 'PPOLoss'
+        importance_ratio, importance_ratio_clipped = value_ops.action_importance_ratio(
+            training_info, self._action_spec, name_scope,
+            self._importance_ratio_clipping, self._log_prob_clipping,
+            self._check_numerics, self._debug_summaries)
         # Pessimistically choose the maximum objective value for clipped and
         # unclipped importance ratios.
         pg_objective = -importance_ratio * advantages
         pg_objective_clipped = -importance_ratio_clipped * advantages
         policy_gradient_loss = tf.maximum(pg_objective, pg_objective_clipped)
 
         if self._debug_summaries and common.should_record_summaries():
-            with tf.name_scope('PPOLoss'):
-                if self._importance_ratio_clipping > 0.0:
-                    clip_fraction = tf.reduce_mean(
-                        input_tensor=tf.cast(
-                            tf.greater(
-                                tf.abs(importance_ratio - 1.0), self.
-                                _importance_ratio_clipping), tf.float32))
-                    tf.summary.scalar('clip_fraction', clip_fraction)
-
-                tf.summary.histogram('action_log_prob', action_log_prob)
-                tf.summary.histogram('action_log_prob_sample',
-                                     sample_action_log_probs)
-                tf.summary.histogram('importance_ratio', importance_ratio)
-                tf.summary.scalar(
-                    'importance_ratio_mean',
-                    tf.reduce_mean(input_tensor=importance_ratio))
-                tf.summary.histogram('importance_ratio_clipped',
-                                     importance_ratio_clipped)
+            with tf.name_scope(name_scope):
```

Comment:
I think using same name_scope at two different places will generate two different name scopes with different suffix by tensorflow.

Reply:
Oops, this sounds scary.  What's the convention here?  Should we leave all debug summaries inside loss classes or is it also Ok to have them in value_ops but need to remember not to do them again in loss classes?

Reply:
Researched a bit, as long as the with tf.name_scope statements are not nested, concatenation of scopes will not happen, see example here:
https://www.tensorflow.org/api_docs/python/tf/Graph#for_example_6

Reply:
Ok, I tested and this seems to be the only way to make it reuse an existing scope.  It's a bit cumbersome, but hopefully not many callers need to invoke this function..